swap out kubectl

mplsgrant opened the pull request:
Took a stab at swapping out kubectl addressing #609

* It appears to *mostly* work; however, something is up with the signet test, and I didn't have a chance to figure out what.

* There was also an oddball issues around `stream` returning python dicts and lists instead of json. I was able to avoid the impacts of this issue.

* Some things remain less tested such as a few namespace items and the chain save-state functionality.

mplsgrant committed:
remove kubectl from stop_scenarios

mplsgrant committed:
remove kubectl from snapshot_bitcoin_datadir

mplsgrant committed:
`auth`: revamp to not use kubectl

Also, provide some logic around overwriting existing entries and also provide a warning when the
namepsace is not set.

mplsgrant committed:
`_logs`: do a container name dance 'bitcoincore'

We need to specify a container name when we have prometheus logging enabled b/c we end up with a
couple containers per pod.

mplsgrant committed:
remove todo from dag_connect

mplsgrant committed:
clarify console text

mplsgrant committed:
`scenarios_test`: use `warnet logs`

I did a drop-in replacement of `kubectl logs` with `warnet logs`.

mplsgrant committed:
`setup`: remove checks and mentions of kubectl

Since we won't be using kubectl, there is no reason for it to be in the setup command.

mplsgrant committed:
`down`: fixup messages in console

mplsgrant committed:
namespaces: remove kubectl

This seems like it needs further testing.

mplsgrant committed:
`get_static_client`: swap out type in return value

mplsgrant committed:
`kexec`: add kexec to replace kubectl exec

mplsgrant committed:
`kubectl_context`: switch up this functional

This function is not used as far as I can tell, and it is not tested. Perhaps consider removing or
integrating somehow into the namespace work if needed.

mplsgrant committed:
`apply_kubernetes_yaml`: remove kubectl

Not sure if we are using this. Does not appear to be tested.

mplsgrant committed:
`delete_namespace`: removed kubectl

mplsgrant committed:
`delete_pod`: appears to work

mplsgrant committed:
`get_default_namespace`: remove kubectl

mplsgrant committed:
`k8s`: fixup imports

mplsgrant committed:
`rpc`: remove kubectl

Had to do an annoying dance because it seems like stream() wants to convert any json it sees into
python primitives. There's gotta be a better way to deal with that, but I came up with crudely
checking if we are getting a list or dict and literal_eval-ing it. Maybe I'm missing something.
Otherwise, seems to work.

mplsgrant committed:
`debug_log`: just swapped it out with warnet logs

mplsgrant committed:
`grep_logs`: minor change; seems to work

mplsgrant committed:
`get_messages`: lots of changes

I made a number of changes to get_messages, not the least of which was swapping out `cat` with
`base64` for the purposes of file transfer. Appears to work.

mplsgrant committed:
`bitcoin`: import fixups

mplsgrant committed:
removed kubectl from setup_minikube.sh

I think we probably should just get rid of this file.

willcl-ark commented:
Am I correct in thinking that all users may still require `kubectl` to be installed, so that the cluster can be configured? Or does the python lib bypass all that too? 

I had always assumed it hinged on `~/.kube/config` to know *where* to do the things it was going to do, but never actually checked it out. Checking the code it seems like there could be alternatives to `load_kube_config()` (in fact, I see [here](https://github.com/kubernetes-client/python/blob/eeeb43cce3c98eb10ae41967138a6d76dbe67ac1/kubernetes/base/config/kube_config.py#L251) references to google auth, which sounds potentially useful to us on the credential side of things?)

Users may also want `kubectl` for any additional k8s applications, such as `k9s`, to work, unless they too can be tricked by dumping a python-generated config (from this `kubernetes` lib) into the user's *~.kube/config* file ;)

If pure python is preferred, then this looks like a big move in the right direction!

pinheadmz commented on src/warnet/bitcoin.py:
```diff
@@ -35,15 +40,37 @@ def rpc(tank: str, method: str, params: str):
+    if resp.startswith("{") or resp.startswith("["):
+        literal = ast.literal_eval(resp)
+        json_string = json.dumps(literal)
+        return json_string
```
the old `exec_run()` function added this option to `stream()` for this: ` _preload_content=False`

overall i think its better to restore the old `exec_run()` function in k8s.py from pre-italy and just call it here.
That's as far as I got on this project: 756df48b97acb1496b2958e12b895a00

mplsgrant replied:
addressed here via _preload_content=False: https://github.com/bitcoin-dev-project/warnet/pull/614/commits/16f869a2a3725ae61f895e153320d6ea18a13add

pinheadmz commented on src/warnet/bitcoin.py:
```diff
@@ -89,8 +117,12 @@ def grep_logs(pattern: str, show_k8s_timestamps: bool, no_sort: bool):
+            namespace=get_default_namespace(),
+            container=container_name,
+            timestamps=True,
+        )
```
can also use the `_log()` internal function I added to get the log

pinheadmz replied:
(or pod_log())

mplsgrant replied:
fixed up here: https://github.com/bitcoin-dev-project/warnet/pull/614/commits/778a3ebab87474086fe39565e579169f49af2465

pinheadmz commented on src/warnet/bitcoin.py:
```diff
@@ -163,21 +195,26 @@ def get_messages(tank_a: str, tank_b: str, chain: str):
     # Get the IP of node_b
-    cmd = f"kubectl get pod {tank_b} -o jsonpath='{{.status.podIP}}'"
-    tank_b_ip = run_command(cmd).strip()
+    tank_b_pod: V1Pod = sclient.read_namespaced_pod(name=tank_b, namespace=get_default_namespace())
```
why not use the get pod helpers in k8s.py?

mplsgrant replied:
fixed up here: https://github.com/bitcoin-dev-project/warnet/pull/614/commits/6d8ffa091089dfadca4fff2e61331c1e8d109f66
and here: https://github.com/bitcoin-dev-project/warnet/pull/614/commits/e84416bc72411f2dbf00615486428260a5085824

pinheadmz commented on src/warnet/control.py:
```diff
@@ -232,52 +244,61 @@ def run(scenario_file: str, additional_args: tuple[str]):
-            print(f"Please consider waiting for the pod to become available. Encountered: {e}")
-    else:
-        pass  # cancelled by user
+    container_name = "bitcoincore" if pod_name.startswith("tank") else None
```
we can't ever rely on tank names conforming to any kind of convention

mplsgrant replied:
fixed up here: https://github.com/bitcoin-dev-project/warnet/pull/614/commits/04be184b91f8d3bf370532add51e80d47e28bd2f

pinheadmz commented on src/warnet/k8s.py:
```diff
@@ -104,29 +141,54 @@ def apply_kubernetes_yaml_obj(yaml_obj: str) -> None:
-    return stream_command(command)
+def get_default_namespace() -> str:
+    with open(KUBECONFIG) as file:
+        kubeconfig_data = yaml.safe_load(file)
```
will this file always exist? is this the only way to get the namespace?

mplsgrant replied:
I will need to revisit this.

mplsgrant replied:
I replaced open(KUBECONFIG) instances with two functions that handle errors: https://github.com/bitcoin-dev-project/warnet/pull/614/commits/b1365d3160ec51925923c6b0173ff502075d40d1

Regarding getting the namespace, I do believe that the namespace stored in the KUBECONFIG file is the source of truth for the namespace for user OR our default namespace which is "warnet" right now. We talked about changing that on the call to "default". I

pinheadmz commented:
Still a few mentions `git grep kubectl`, but mostly just docs (should update) and bash scripts (do we still need?). The port forward command will be replaced by ingress #607

mplsgrant committed:
`pod_log`: add type; add timestamp option

mplsgrant committed:
`grep_logs`: fixup and use pod_log instead

mplsgrant committed:
`_rpc`: fix goofy json parsing workaround

I went with the _preload_content=False approach

mplsgrant committed:
`k8s`: add get_pod and get_service

mplsgrant committed:
`bitcoin`: use get_pod and get_service

mplsgrant committed:
`_logs`: do not rely on container name

Instead rely on the position of the container.

Position Rule: we get logs from the first container as reported by V1Pod's spec.container metadata.

mplsgrant committed:
clarify default namespace

mplsgrant committed:
`auth`: create kubeconfig from credential file

mplsgrant commented:
> Am I correct in thinking that all users may still require `kubectl` to be installed, so that the cluster can be configured? Or does the python lib bypass all that too?

@willcl-ark  The python lib bypasses this. It currently relies on KUBECONFIG which is genreally at ~/.kube/config

> it seems like there could be alternatives to load_kube_config()

I think the KubeConfigLoader provides access to KUBECONFIG, but it can also do config files in custom locations (maybe in memory, too?). Seems like the google credentials thing might be an easy win to modify KUBECONFIG with the correct google plugin data, but I have not had a chance to look into that yet.

> Users may also want kubectl for any additional k8s applications, such as k9s

I don't think k9s or ktop requires kubectl, but I do think we should guide users towards downloading it b/c it is so useful.

Regarding setting the kubeconfig file, right now we rely on minikube to do that. I have also updated `auth` to create a KUBECONFIG file from a credential file if the user does not have an existing one:  https://github.com/bitcoin-dev-project/warnet/pull/614/commits/c66648081ac25eb11c0c9336b52ec910fb6e9434

mplsgrant commented:
It appears that signet remains an issue. Will need to revisit.

mplsgrant committed:
fix colorization of log output

mplsgrant committed:
clean up imports

mplsgrant committed:
`control`: catch None type; fix coloring

mplsgrant committed:
`test_base`: use shlex to handle cmd splitting

shlex appears to handle escape sequences nicely.

mplsgrant committed:
`k8s`: add a K8sError

I figure we should add an error type for the k8s file, that way we can try...except everything from
that file using just one error type.

mplsgrant committed:
`k8s`: update type in create_kubernetes_object

mplsgrant committed:
`k8s`: add open and write fns for KUBECONFIG

mplsgrant committed:
`k8s`: update namespace and context checking fns

mplsgrant committed:
`users`: improve auth fn

Incorporates better error handling

mplsgrant commented:
> It appears that signet remains an issue. Will need to revisit.

I fixed up the signet test issue: https://github.com/bitcoin-dev-project/warnet/pull/614/commits/02562b216fd26ad201f5929e57a628fae2ccd4ac

It uses shlex to split up lines. Allegedly, shlex is command-line-escape-character aware.

mplsgrant committed:
`docs`: minor doc updates

mplsgrant committed:
`bitcoin`: remove debug_log

mplsgrant committed:
`control`: add explicit ns to run --debug

mplsgrant committed:
`control`: remove contrived get_default_namespace

This use of get_default_namespace relies on the brittle idea that get_pods will always use
get_default_namespace. Currently, get_pods currently does, in fact, use get_default_namespace.
However, I don't like that we use get_default_namespace out of band relative to get_pods. I'd prefer
that we do not mention namespaces at all in our message to users unless we can prove it.

mplsgrant committed:
`k8s`: use our internal get_pod fn

mplsgrant committed:
`k8s`: use default namespace for `edges()`

mplsgrant committed:
DEFAULT_NAMESPACE: set it to "default"

mplsgrant committed:
`k8s`: use default namespace in apply fn

mplsgrant committed:
`debug_log`: bring it back

mplsgrant committed:
k8s: add `continue` to `wait_for_init`

mplsgrant committed:
make ruff happy

mplsgrant committed:
add constants as they appear in the helm charts

mplsgrant committed:
use constants in the `log` function

mplsgrant committed:
use constant in the bitcoin `rpc` function
